core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StreamPropertyDerivations.java

```diff
@@ -338,6 +338,10 @@ public StreamProperties visitExchange(ExchangeNode node, List<StreamProperties>
                     if (node.getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_ARBITRARY_DISTRIBUTION)) {
                         return new StreamProperties(FIXED, Optional.empty(), false);
                     }
+                    // empty arguments list means the bucketing function is effectively constant (1 bucket)
+                    if (node.getPartitioningScheme().getPartitioning().getArguments().isEmpty()) {
```

Comment:
I'm not sure this always works.
For example `LocalExecutionPlanner.Visitor#visitTableWriter` sets:
```
context.setDriverInstanceCount(getTaskWriterCount(session));
```
while `LocalExecutionPlanner.Visitor#createMergeSource` sets:
```
context.setDriverInstanceCount(1);
```
This would fail if such operators are part of same pipeline. We use `StreamProperties` to add local exchanges which split single large pipeline into separate ones.

Perhaps this works in your use case, but I think this breaks definition of stream properties. You need exactly one stream (set via `setDriverInstanceCount(1)`) to be able to use `singleStream`.

Reply:
I think it's also wrong because connector could return some random partitioning even if there are no arguments.
I would just skip property changes in this PR

Reply:
there is unwritten, pre-existing assumption that partitioning must be deterministic. it doesn't seem to make sense otherwise

Reply:
It doesn't have to be deterministic, see `io.trino.sql.planner.SystemPartitioningHandle.SystemPartitionFunction.RoundRobinBucketFunction`

Reply:
> This would fail if such operators are part of same pipeline.

Can you think of any scenario that would actually create such a pipeline?

> I would just skip property changes in this PR

Unfortunately changing this property is the core of this pr - at least I don't see any other ways to do this.

Reply:
> Unfortunately changing this property is the core of this pr - at least I don't see any other ways to do this.

Why it's in the core? You need one effective writer and empty partitioning provides you with that.

Reply:
Do I understand correctly - you are saying that returning `new StreamProperties(FIXED, Optional.empty(), false);` would be fine?

Reply:
Yes. I don't think `StreamPropertyDerivations.java` needs to be changed

Reply:
Perhaps this check is no longer valid (it was introduced in 2016). Does anything breaks without it?

Reply:
> Perhaps this check is no longer valid (it was introduced in 2016). Does anything breaks without it?

Hmm I didn't consider that

Reply:
> Perhaps this check is no longer valid (it was introduced in 2016). Does anything breaks without it?

I am also little confused on how to verify it. Normally I'd rely on tests. However the PR is green in current state and you say it is incorrect so it shows that relying on tests is not sufficient.  Would the tests be sufficient for verification of removal of that check?

Reply:
Yes, running tests should be enough. If something doesn't work, should be easy to fix

Reply:
Ok but as you can see in this PR tests also passed yet you claim this is incorrect why trusting tests then and not now ?

Reply:
I've described why this breaks relationship between properties and local execution planner (https://github.com/trinodb/trino/pull/10460#discussion_r781374496).

I don't think execution model enforces
```
checkArgument(distribution == SINGLE || !this.partitioningColumns.equals(Optional.of(ImmutableList.of())),
                    "Multiple streams must not be partitioned on empty set");
```
(for example on global level it's fine to have non-deterministic partitioning with empty set).
I think that check was just added a bit eagerly.

Reply:
> (for example on global level it's fine to have non-deterministic partitioning with empty set).

that doesn't make sense to me as a feature. why would connector return a partitioning at all?

and, i don't accept this as a design decision. Determinism is an important aspect, and not to be given away without good reason.

Reply:
> and, i don't accept this as a design decision. Determinism is an important aspect, and not to be given away without good reason.

One could imagine partitioning that is based on load (e.g when writing data from one system to the other). It doesn't have to be connector provided (it could be system partitioning, just using SPI)

Reply:
I removed the check